controller/src/bin/xcvradm.rs

```diff
@@ -13,29 +20,145 @@ async fn request_handler(_: SpRpcRequest) -> Result<HostRpcResponse, Error> {
+        value_parser = clap::value_parser!(u64).range(1..=10000)
+    )]
+    retry_interval: u64,
+
+    /// The log-level.
+    #[arg(
+        short,
+        long,
+        default_value_t = Level::Info,
+        value_parser = parse_log_level
+    )]
+    log_level: Level,
+}
+
+#[derive(Subcommand)]
+enum Cmd {
+    /// Return the status of the addressed modules, such as presence, power
+    /// enable, and power mode.
+    Status,
+
+    /// Read the lower page of a set of transceiver modules.
+    ReadLower { offset: u8, len: u8 },
+}
+
 #[tokio::main]
 async fn main() {
+    let args = Args::parse();
+    let config = Config {
+        address: args.address,
+        interface: args.interface,
+        peer: args
+            .peer
+            .unwrap_or(Ipv6Addr::from(transceiver_messages::ADDR)),
+        n_retries: args.n_retries,
+        retry_interval: Duration::from_millis(args.retry_interval),
+    };
+
     let decorator = slog_term::TermDecorator::new().build();
     let drain = slog_term::FullFormat::new(decorator).build().fuse();
     let drain = slog_async::Async::new(drain).build().fuse();
+    let drain = slog::LevelFilter::new(drain, args.log_level).fuse();
     let log = slog::Logger::root(drain, slog::o!());
 
     let request_handler_log = log.new(slog::o!("name" => "request_handler"));
     let handler = move |message| {
         slog::info!(request_handler_log, "received sp request"; "message" => ?message);
         async move { request_handler(message).await }
     };
-    let controller = Controller::new(log.clone(), handler).await.unwrap();
-    let read = MemoryRead::new(sff8636::Page::Lower, 0, 4).unwrap();
-    let data = controller
-        .read(
-            ModuleId {
-                fpga_id: 0,
-                ports: PortMask(0b11),
-            },
-            read,
-        )
-        .await
-        .unwrap();
-    slog::info!(log, "data: {:?}", data);
+
+    let ports = args
+        .transceivers
+        .map(|list| PortMask::from_indices(&parse_xvcr_list(&list).unwrap()).unwrap())
+        .unwrap_or_else(|| PortMask::all());
+    let modules = ModuleId {
+        fpga_id: args.fpga_id,
+        ports,
+    };
+    let controller = Controller::new(config, log.clone(), handler).await.unwrap();
+
+    match args.cmd {
+        Cmd::Status => {
+            let status = controller.status(modules).await.unwrap();
+            print_module_status(modules, status);
+        }
+        Cmd::ReadLower { offset, len } => {
+            let read = MemoryRead::new(sff8636::Page::Lower, offset, len).unwrap();
+            let data = controller.read(modules, read).await.unwrap();
+            print_read_data(modules, data);
+        }
+    }
+}
+
+fn print_module_status(modules: ModuleId, status: Vec<Status>) {
+    println!("FPGA\tPort\tStatus");
+    for (port, status) in modules.ports.to_indices().zip(status.into_iter()) {
+        println!("{}\t{port}\t{status:?}", modules.fpga_id);
```

Comment:
In Humility tables, we use fixed-width format strings (rather than tabs, which can vary from terminal to terminal):
```suggestion
        println!("{:4>} {port:4>} {status:?}", modules.fpga_id);
```
(this comment is generally applicable to all of the `print_*` functions here)

Reply:
Nice, that's a good idea thanks!